### ISSUE TITLE
maximum_number_of_concurrent_rule_engine_server_processes seems not to work

### ISSUE THREAD

**Issue by bh9** [irods/irods#5482]
### iRODS Version, OS and Version
4.2.7, Ubuntu 18.04 running 16.04's packages

### What did you try to do?
While indexing, we observed an oom reap of the ReServer with `"maximum_number_of_concurrent_rule_engine_server_processes": 4,` in the config. 

### Expected behavior
Our understanding of maximum_number_of_concurrent_rule_engine_server_processes was that it would limit the number of child processes ReServer would have to 4

### Observed behavior (including steps to reproduce, if applicable)
ReServer had 384 children when it was reaped. 

oom reap:
```
Mar 12 03:18:03 irods-seq-indexing kernel: [14830091.770115] Out of memory: Kill process 26279 (irodsReServer) score 122 or sacrifice child
Mar 12 03:18:03 irods-seq-indexing kernel: [14830091.772205] Killed process 26279 (irodsReServer) total-vm:1986364kB, anon-rss:1448448kB, file-rss:1696kB, shmem-rss:0kB
Mar 12 03:18:03 irods-seq-indexing kernel: [14830091.860792] oom_reaper: reaped process 26279 (irodsReServer), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB
```

All 384 children logged to rodsLog when they died, each of the form:
```
Mar 12 03:18:04 pid:6845 remote addresses: 172.17.72.39, 172.27.208.167 ERROR: [-]      /irods/server/core/src/rsApiHandler.cpp:540:int readAndProcClientMsg(rsComm_t *, int) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [failed to call 'read header']
        [-]     /irods/lib/core/src/sockComm.cpp:201:irods::error readMsgHeader(irods::network_object_ptr, msgHeader_t *, struct timeval *) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [failed to call 'read header']
                [-]     /irods/plugins/network/ssl/libssl.cpp:571:irods::error ssl_read_msg_header(irods::plugin_context &, void *, struct timeval *) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [read 0 expected 4]

Mar 12 03:18:04 pid:6845 remote addresses: 172.17.72.39 ERROR: [-]      /irods/server/core/src/rodsAgent.cpp:542:int runIrodsAgentFactory(sockaddr_un) :  status [SSL_SHUTDOWN_ERROR]  errno [] -- message [failed to call 'agent stop']
        [-]     /irods/lib/core/src/sockComm.cpp:163:irods::error sockAgentStop(irods::network_object_ptr) :  status [SSL_SHUTDOWN_ERROR]  errno [] -- message [failed to call 'agent stop']
                [-]     /irods/plugins/network/ssl/libssl.cpp:942:irods::error ssl_agent_stop(irods::plugin_context &) :  status [SSL_SHUTDOWN_ERROR]  errno [] -- message [error completing shutdown of SSL connection]

Mar 12 03:18:05 pid:26276  ERROR: Agent process [6845] exited with status [64]
```

**Comment by bh9** [comment]
On closer inspection, these would appear to be clients, not children

**Comment by trel** [comment]
Whew - okay, so, it appears from the log...
pid 26276 is the `irodsServer` and
pid 6845 is the `irodsAgent` spun up to service the that delayed rule execution.

The OOM is working as designed, then?    Is there still a bug?   A leak in the `irodsReServer`?

**Comment by bh9** [comment]
Sure, oom is working as expected. I guess what matters for us is how we can stop the rule engine from creating quite so many clients. There are currently only ~200k rules, so I think the issue is not the number of rules in the queue, but the number of active clients
It is interesting this happened just after it started indexing new documents (well, ~1 million documents later but that's 0.5% of the zone)

**Comment by bh9** [comment]
This is just a guess:
It would make sense that new documents would take longer than existing documents, because elasitcsearch has to actually write them to disk. Given this, if we assume that the number of clients is at the limit of a bottleneck while writing existing documents, then we would expect the number of clients to rise when we increase their runtime by writing new documents, without reducing the rate at which they are produced. If these are left to build up, then eventually the host exhausts itself of memory and picks something to oom, and what better than the ReServer which is holding the state of the full collection scan

**Comment by alanking** [comment]
This seems very similar to the issue described in #4616. The "thundering herd" problem has been resolved as of 4.2.8 with connection pooling to make its return on #4735's resolution.

**Comment by trel** [comment]
I think this is a duplicate.    Will leave open for a bit, but putting in 4.2.9 so we don't forget.

**Comment by korydraughn** [comment]
I believe this can be closed. The delay server no longer grows child processes exponentially.

See https://github.com/irods/irods/issues/4616.

**Comment by trel** [comment]
closing - was addressed in 4.2.8 - if we find this again, we'll open a new issue.
